File: 42-trapping-rain-water/trapping-rain-water.ts
function trap(height: number[]): number {
    let l = 0;
    let r = height.length - 1;
    let lMax = height[0];
    let rMax = height[height.length - 1];

    let count = 0;

    while (l < r) {
        if (rMax <= lMax) {
            count += rMax - height[r--];
            rMax = Math.max(rMax, height[r]);
        }
        else {
            count += lMax - height[l++];
            lMax = Math.max(lMax, height[l]);
        }
    }

    return count;
};

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int leftBoundary = height[left];
        int rightBoundary = height[right];
        int totalWaterCount = 0;
        while(left < right){
            leftBoundary = Math.max(leftBoundary , height[left]);
            rightBoundary = Math.max(rightBoundary , height[right]);
            // System.out.printf("%d , %d , %d , %d \n", left , right , leftBoundary , rightBoundary);
            if(leftBoundary <= rightBoundary){
                totalWaterCount += Math.min(leftBoundary , rightBoundary) - height[left];
                left ++;
            }
            else{
                totalWaterCount += Math.min(leftBoundary , rightBoundary) - height[right];
                right --; 
            }
        }
        return totalWaterCount;
    }
}